### ISSUE TITLE
AFQMC xxx_buffer_generators are global

### ISSUE THREAD

**Issue by PDoakORNL** [QMCPACK/qmcpack#2433]
**Is your feature request related to a problem? Please describe.**
A great deal of trouble and annoyance is still being causes in the realspace QMCPACK because the decision was made to make many important objects globals. It's actually worth it to make an explicit decision about such practices early in a project.

Currently std::shared_ptr is used to create set of pointers to global xxx_buffer_generators.
I don't think this is better practice that simply using a raw pointer since you should be able to reason about the order of your object destruction and this implies users of this shared_ptr may need the global objects destruction delayed past the end of main.

I would argue they should/must be owned by a particular scope and not the "runtime." The consumer of the allocator should be able to control and document as code what the scope of the allocator is.

Global objects cause issue both for unit testing and for concurrent programming.

**Describe the solution you'd like**
Removal reliance on global shared_ptrs.  Choose a scope and lifetime for the allocators.

**Comment by mmorale3** [comment]
I thought  about this for a long time and I decided to stick with global objects.
A new PR changes slightly the access pattern to the allocator generators, I now use a 
monostate-type pattern, but the objects are still global. 
The changes needed to provide a scoped access to these objects would be too great and
I'm not willing to go that route. I need this allocators at both the top level of the code and 
at the lowest level during dispatch calls to cuda libraries (cutensor, cublas and cusparse now
all require workspace, which is a pain to deal with).
Anyway, just letting you know that I made a decision and that it is addressed in a recent PR.
